core/src/main/java/org/apache/iceberg/rest/ErrorHandlers.java

```diff
@@ -128,6 +157,39 @@ public static Consumer<ErrorResponse> defaultErrorHandler() {
       }
 
       throw new RESTException("Unable to process: %s", error.message());
-    };
+    }
+  }
+
+  private static class OAuthErrorHandler extends ErrorHandler {
+    private static final ErrorHandler INSTANCE = new OAuthErrorHandler();
+    private static final String SERVER_ERROR = "server_error";
+
+    @Override
+    public ErrorResponse parseResponse(int code, String json) {
+      if (code == 400 || code == 401) {
+        return OAuthErrorResponseParser.fromJson(code, json);
+      }
+      return ErrorResponse.builder()
+          .responseCode(code)
+          .withType(SERVER_ERROR)
+          .withMessage(json)
+          .build();
+    }
+
+    @Override
+    public void accept(ErrorResponse error) {
+      switch (error.type()) {
+        case OAuth2Properties.INVALID_CLIENT_ERROR:
+          throw new NotAuthorizedException("Not authorized: %s", error.message());
+        case OAuth2Properties.INVALID_REQUEST_ERROR:
+        case OAuth2Properties.INVALID_GRANT_ERROR:
+        case OAuth2Properties.UNAUTHORIZED_CLIENT_ERROR:
+        case OAuth2Properties.UNSUPPORTED_GRANT_TYPE_ERROR:
+        case OAuth2Properties.INVALID_SCOPE_ERROR:
+          throw new BadRequestException("Malformed request: %s", error.message());
```

Comment:
Should we also include the error type in the exception?

Reply:
Yes, that makes sense, I will add that.